lib/workflow/stores/mongo.js

```diff
@@ -639,22 +639,39 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
         assert.uuid(data.taskId, 'data.taskId');
         assert.string(data.state, 'data.state');
 
-        var query = {
+        var queryDependency = {
             graphId: data.graphId,
-            reachable: true
+            reachable: true,
         };
-        query['dependencies.' + data.taskId] = {
+        queryDependency['dependencies.' + data.taskId] = {
             $in: [data.state, Constants.Task.States.Finished]
         };
-        var update = {
+        
+        var updateDependency = {
+            $unset: {}
+        };
+        updateDependency.$unset['dependencies.' + data.taskId] = '';
+
+        var queryAnyOf = {
+            graphId: data.graphId,
+            reachable: true,
+        };
+        queryAnyOf['dependencies.anyOf.' + data.taskId] = {
+            $in: [data.state, Constants.Task.States.Finished]
+        };
+        var updateAnyOf = {
             $unset: {}
         };
-        update.$unset['dependencies.' + data.taskId] = '';
+        updateAnyOf.$unset["dependencies.anyOf"] = "";
+
         var options = {
             multi: true
         };
-
-        return waterline.taskdependencies.updateMongo(query, update, options);
+        
+        return Promise.all([
```

Comment:
@yyscamper @zhangy42 @pscharla  I think concurrent execution is fine because writes in mongo are atomic and it doesn't look like the two updates could try to unset the same same thing since they query/unset dependencies.{someTaskId} and dependencies.anyOf.{someTaskId} respectively.